### PR TITLE
[SUCE-77] pali  lookup

### PR DESCRIPTION
**Pull request by gradam**

### CODE REVIEW

File: client/elements/text/sc-simple-text.html [hunk 1]
```diff
@@ -259,6 +299,153 @@
...
+            let node;
+            for(let i=0;i<nodes.length;i++){
+                node = nodes[i];
+                this._addSpanToNode(node);
+                if(node.childNodes){
+                    this._recurseDomChildren(node, output);
+                }
+            }
+        }
+        _addSpanToNode(node){
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: ditto

File: client/elements/text/sc-simple-text.html [hunk 2]
```diff
@@ -259,6 +299,153 @@
...
+            if(node.nodeType != 3) return;
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: !==

File: client/elements/text/sc-simple-text.html [hunk 3]
```diff
@@ -259,6 +299,153 @@
...
+            let tt = node.data;
+            let strArr = tt.split(" ");
+            let str = "";
+            for( let i = 0 ; i < strArr.length; i++) if (strArr[i]){
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: ditto

File: client/elements/text/sc-simple-text.html [hunk 4]
```diff
@@ -259,6 +299,153 @@
...
+                str += 'spfrnt' + strArr[i] + 'spback ';
+            }
+            node.data = str;
+        }
+    
+        _addTooltip(v, tooltip, paliLookup, states, e) {
+            let timeToWait = e.sourceCapabilities.firesTouchEvents ? 10 : 100;
+            let explenation = paliLookup.lookupWord(v.innerText);
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: ditto

File: client/elements/text/sc-simple-text.html [hunk 5]
```diff
@@ -259,6 +299,153 @@
...
+                tooltip.animationDelay = 0;
+                tooltip.manualMode = true;
+                tooltip.fitToVisibleBounds = true;
+                tooltip.offset = 0;
+                let states = {isWordHovered: false, 
+                            isTooltipHovered: false,
+                            isTooltipShowed: false};
+                v.addEventListener('mouseover', (e) => {
+                    if(this.isPaliLookupEnabled && this.isLookupAllowed){
+                        v.style.color = '#43A047'; // It can not be in class because of some strange bug in some cases.
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: ditto

File: client/elements/text/sc-simple-text.html [hunk 6]
```diff
@@ -259,6 +299,153 @@
...
+            }
+            
+            let paliLookup = self.shadowRoot.querySelector('#pali_lookup');
+            textContainer.querySelectorAll('.word').forEach((v) => {
+                let tooltip = document.createElement('paper-tooltip');
+                tooltip.animationDelay = 0;
+                tooltip.manualMode = true;
+                tooltip.fitToVisibleBounds = true;
+                tooltip.offset = 0;
+                let states = {isWordHovered: false,
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: formatting is wrong here

File: client/elements/text/sc-simple-text.html [hunk 7]
```diff
@@ -259,6 +299,153 @@
...
+        _paliLookupStateChanged(){
+            if(this.hidden){
+                return;
+            }
+            if(this.isPaliLookupEnabled){
+            if(!this.spansGenerated){
+                if(this.lang === 'pli' || this.lang === 'lzh') {
+                    this._wordsIntoSpans('.sutta');
+                } 
+            }
```
Comment (TrebuhD) [inline client/elements/text/sc-simple-text.html]: indentation

File: client/redux-store.html [hunk 8]
```diff
@@ -63,6 +65,11 @@
             case 'CHOOSE_PALI_TEXT_SCRIPT':
                 return Object.assign({}, state,
                     { textOptions: Object.assign({}, state.textOptions, { script: action.script }) });
+            case 'ACTIVATE_PALI_LOOKUP':
+                return Object.assign({}, state.paliLookup,
+                    { textOptions: Object.assign({}, state.textOptions, 
+                                            {paliLookupActivated: action.paliLookupActivated,
```
Comment (TrebuhD) [inline client/redux-store.html]: whitespace

File: client/redux-store.html [hunk 9]
```diff
@@ -63,6 +65,11 @@
             case 'CHOOSE_PALI_TEXT_SCRIPT':
                 return Object.assign({}, state,
                     { textOptions: Object.assign({}, state.textOptions, { script: action.script }) });
+            case 'ACTIVATE_PALI_LOOKUP':
+                return Object.assign({}, state.paliLookup,
```
Comment (TrebuhD) [inline client/redux-store.html]: `state.paliLookup` doesn't exist anymore in the`initialState` object

File: client/elements/lookups/sc-pli.html [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+          }
+
+          _toHtml(allMatches, original) {
+            let out = "";
+              if (allMatches.length === 0) {
+                  allMatches.push({ "base": original, "meaning": "?" });
+              }
+              if (this.isTi) allMatches.push({ "base": "iti", "meaning": "endquote" });
+              for (let match of allMatches) {
+                  let href = "http://suttacentral.net/define/" + match.base;
```
[the remark this answers is not in the document]
Comment (gradam) [reply client/elements/lookups/sc-pli.html]: yes. It points to the old website for now

File: client/elements/text/sc-segmented-text.html [hunk 11]
```diff
@@ -411,6 +454,155 @@
...
+                    self._removeTooltip(v, tooltip, states, e);
+                });
+            })
+          }
+          _paliLookupStateChanged(){
+              if(this.hidden){
+                  return;
+              }
+              if(this.isPaliLookupEnabled){
+                if(!this.spansGenerated){
```
[the remark this answers is not in the document]
Comment (gradam) [reply client/elements/text/sc-segmented-text.html]: done

**Pull request merged**